lib/Conversion/GPUToROCM/ConvertKernelFuncToHSACO.cpp

```diff
@@ -0,0 +1,409 @@
+#include "llvm/Support/TargetRegistry.h"
+#include "llvm/Support/TargetSelect.h"
+#include "llvm/Target/TargetMachine.h"
+#include "llvm/Transforms/IPO/Internalize.h"
+
+#include <fstream>
+#include <iostream>
+#include <string>
+
+using namespace mlir;
+
+namespace {
+
+/// A pass converting tagged kernel functions to HSA Code Object blobs.
+///
+/// If tagged as a kernel module, each contained function is translated to ROCDL
+/// IR, which is then compiled using the llvm AMDGPU backend to generate the GPU
+/// binary code (i.e. the HSACO file). The HSACO binary blob is attached as an
+/// attribute to the function and the function body is erased.
+class GpuKernelToHSACOPass : public ModulePass<GpuKernelToHSACOPass> {
+public:
+  GpuKernelToHSACOPass(rocm::HSACOGeneratorConfig hsacoGeneratorConfig =
+                           rocm::HSACOGeneratorConfig(/*isTestMode*/ true))
+      : config(hsacoGeneratorConfig) {}
+
+  // Run the dialect converter on the module.
+  void runOnModule() override {
+
+    // Nothing to do if this module does not contain the "gpu.kernel_module"
+    // attribute,
+    // which is used to mark the (nested) modules created to house the GPU
+    // kernel functions
+    if (!getModule().getAttrOfType<UnitAttr>(
+            gpu::GPUDialect::getKernelModuleAttrName())) {
+      return;
+    }
+
+    // This is a module containing a GPU kernel function, we have work to do!
+
+    // Make sure the AMDGPU target is initialized.
+    LLVMInitializeAMDGPUTarget();
+    LLVMInitializeAMDGPUTargetInfo();
+    LLVMInitializeAMDGPUTargetMC();
+    LLVMInitializeAMDGPUAsmPrinter();
+
+    auto llvmModule = translateModuleToROCDLIR(getModule());
+    if (!llvmModule) {
+      return signalPassFailure();
+    }
+
+    for (auto function : getModule().getOps<FuncOp>()) {
+      if (!gpu::GPUDialect::isKernel(function)) {
+        continue;
+      }
+
+      if (failed(translateGpuKernelToHSACOAnnotation(*llvmModule, function)))
+        signalPassFailure();
+    }
+
+    // dump the MLIR
+    if (getenv("MLIR_ROCM_DUMP_AFTER_CONV_GPU_KERNEL")) {
+      std::cerr << "MLIR - After call to convert GPU Kernel" << std::endl;
+      getModule().dump();
+    }
+  }
+
+private:
+  LogicalResult translateGpuKernelToHSACOAnnotation(llvm::Module &llvmModule,
+                                                    FuncOp &function);
+
+  OwnedHSACO convertModuleToHSACO(llvm::Module &llvmModule, FuncOp &function);
+
+  OwnedHSACO emitModuleToHSACO(llvm::Module &llvmModule,
+                               llvm::TargetMachine &targetMachine);
+
+  OwnedHSACO emitModuleToHSACOForTesting(llvm::Module &llvmModule,
+                                         FuncOp &function);
+
+  rocm::HSACOGeneratorConfig config;
+};
+
+// get the "-mcpu" option string corresponding to the given AMDGPU version enum
+std::string GetMcpuOptionString(rocm::AMDGPUVersion v) {
+  switch (v) {
+  case rocm::AMDGPUVersion::GFX900:
+    return "gfx900";
+    break;
```

Comment:
The 'break's aren't necessary on any of these.

Reply:
will remove them in the next push